### PR TITLE
8271959: [lworld] Convert Optional and VBS classes to value class

### PR DESCRIPTION
**Pull request by RogerRiggs**
Make copies of classes to be made value classes in src/java.base/valueclasses/classes 
Modify CompileJavaModules.gmk to trigger builds of values classes and construct a .jar for each module.
Add the jar files to $JAVA_HOME/lib/valueclasses/\<module\>-valueclasses.jar 

Modify hotspot arguments.cpp to scan for patch jar files when --enable-preview and -XX:+EnableValhalla.
For each jar, the equivalent of --patch-module <module>=<path-to-jar> is added and the system properties `jdk.module.patch.<n>` include jar file paths.

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Issue
 * [JDK-8271959](https://bugs.openjdk.org/browse/JDK-8271959): [lworld] Convert Optional and VBS classes to value class

### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - no project role) ⚠️ Review applies to [8ed83988](https://git.openjdk.org/valhalla/pull/816/files/8ed839880676d3474b19efc833731fc1923b6b96)
 * [Mandy Chung](https://openjdk.org/census#mchung) (@mlchung - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/valhalla pull/816/head:pull/816` \
`$ git checkout pull/816`

Update a local copy of the PR: \
`$ git checkout pull/816` \
`$ git pull https://git.openjdk.org/valhalla pull/816/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 816`

View PR using the GUI difftool: \
`$ git pr show -t 816`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/valhalla/pull/816.diff">https://git.openjdk.org/valhalla/pull/816.diff</a>

</details>

### CODE REVIEW

**Comment by vova7878** [comment]
Are these changes part of JEP 402? Will java.Integer etc. replaced in a similar way?

**Comment by RogerRiggs** [comment]
> Are these changes part of JEP 402? Will java.lang.Integer etc. replaced in a similar way?

I think that's unlikely, at least as sketched here.  The wrapper types are used before the module system is initialized.
Likely, something more specific and low level will be needed for those.

**Comment by RogerRiggs** [comment]
The initial set of value classes (listed in `make/modules/java.base/gensrc/GensrcValueClasses.gmk`) are generated from the regular versions of those classes with the `value` keyword inserted in the class declaration.

    java/util/Optional.java \
    java/util/OptionalInt.java \
    java/util/OptionalLong.java \
    java/util/OptionalDouble.java \
    java/time/LocalDate.java \
    java/time/LocalDateTime.java \
    java/time/LocalTime.java \
    java/time/Duration.java \
    java/time/Instant.java \
    java/time/MonthDay.java \
    java/time/ZonedDateTime.java \
    java/time/OffsetDateTime.java \
    java/time/OffsetTime.java \
    java/time/YearMonth.java \
    java/time/Year.java \
    java/time/Period.java \

File: src/hotspot/share/runtime/arguments.cpp [hunk 1]
```diff
@@ -2218,18 +2218,79 @@ int Arguments::process_patch_mod_option(const char* patch_mod_tail, bool* patch_
       memcpy(module_name, patch_mod_tail, module_len);
       *(module_name + module_len) = '\0';
       // The path piece begins one past the module_equal sign
-      add_patch_mod_prefix(module_name, module_equal + 1, patch_mod_javabase);
+      add_patch_mod_prefix(module_name, module_equal + 1, false);
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/arguments.cpp:2221]: `patch_mod_javabase` seems unused here now. Not obvious that this flag is being tracked properly with the new changes.

File: src/hotspot/share/runtime/arguments.cpp [hunk 2]
```diff
@@ -2218,18 +2218,79 @@ int Arguments::process_patch_mod_option(const char* patch_mod_tail, bool* patch_
...
+  // If --enable-preview and EnableValhalla is true, each module may have value classes that
+  // are to be patched into the module.
+  // For each <module>-valueclasses.jar in <JAVA_HOME>/lib/valueclasses/
+  // appends the equivalent of --patch-module <module>=<JAVA_HOME>/lib/valueclasses/<module>-valueclasses.jar
+  if (enable_preview() && EnableValhalla) {
+    char * valueclasses_dir = AllocateHeap(JVM_MAXPATHLEN, mtArguments);
+    const char * fileSep = os::file_separator();
+
+    jio_snprintf(valueclasses_dir, JVM_MAXPATHLEN, "%s%slib%s" VALUECLASS_STR "%s",
+        Arguments::get_java_home(), fileSep, fileSep, fileSep);
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/arguments.cpp:2246]: Nit: please indent continuation line so 'A' and 'v' line up.

File: src/hotspot/share/runtime/arguments.cpp [hunk 3]
```diff
@@ -2218,18 +2218,79 @@ int Arguments::process_patch_mod_option(const char* patch_mod_tail, bool* patch_
...
+        int len = (int)strlen(entry->d_name) - (sizeof(VALUECLASS_JAR) - 1);
+        if (len <= 0 || strcmp(&entry->d_name[len], VALUECLASS_JAR) != 0) {
+          continue;         // too short or not the expected suffix
+        }
+
+        strcpy(module_name, entry->d_name);
+        module_name[len] = '\0';     // truncate to just module-name
+
+        jio_snprintf(path, JVM_MAXPATHLEN, "%s%s", valueclasses_dir, &entry->d_name);
+        add_patch_mod_prefix(module_name, path, true);
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/arguments.cpp:2263]: Please document the `true` e.g. `true /* append */` (similar for the false case elsewhere)

**Review by dholmes-ora — COMMENTED** [review]
Mostly seems reasonable - though I'm not that familiar with the existing module patching logic. One issue with tracking `patch_mod_javabase`.

File: src/hotspot/share/runtime/arguments.cpp (hunk 1 above)
Comment (RogerRiggs) [reply src/hotspot/share/runtime/arguments.cpp]: Patch_mod_javabase is obsolete and is to be removed.
Each --patch-module is added to an array as the arguments are parsed.
After the arguments are parsed the array is checked to see if any modules have been patched.
That is sufficient for the decision to disable CDS.

File: src/hotspot/share/runtime/arguments.cpp [hunk 4]
```diff
@@ -3191,8 +3189,8 @@ jint Arguments::finalize_vm_init_args(bool patch_mod_javabase) {
     }
   }
 
-  if (UseSharedSpaces && patch_mod_javabase) {
-    no_shared_spaces("CDS is disabled when " JAVA_BASE_NAME " module is patched.");
+  if (UseSharedSpaces && _patch_mod_prefix != nullptr) {
+    no_shared_spaces("CDS is disabled when any module is patched.");
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/arguments.cpp:3193]: But you are now disabling CDS if any module is patched, when the constraint AFAIK is only on the base module.

Comment (RogerRiggs) [reply src/hotspot/share/runtime/arguments.cpp]: Corrected to restore previous behavior and different messages between patching java.base and any other module.

**Comment by RogerRiggs** [comment]
@erikj79 When you have time can you look at the build changes. Thanks

**Comment by mlchung** [comment]
Hi @mlchung, thanks for making a comment in an OpenJDK project!

All comments and discussions in the OpenJDK Community must be made available under the OpenJDK [Terms of Use](https://openjdk.java.net/legal/tou/). If you already are an OpenJDK [Author](https://openjdk.java.net/bylaws#author), [Committer](https://openjdk.java.net/bylaws#committer) or [Reviewer](https://openjdk.java.net/bylaws#reviewer), please click [here](https://bugs.openjdk.java.net/secure/CreateIssue.jspa?pid=11300&issuetype=1) to open a new issue so that we can record that fact. Please Use "Add GitHub user mlchung" for the summary.

If you are not an OpenJDK Author, Committer or Reviewer, simply check the box below to accept the OpenJDK Terms of Use for your comments.

 - [ ]  I agree to the [OpenJDK Terms of Use](https://openjdk.java.net/legal/tou/) for all comments I make in a project in the [OpenJDK GitHub organization](https://github.com/openjdk).

Your comment will be automatically restored once you have accepted the OpenJDK [Terms of Use](https://openjdk.java.net/legal/tou/).

File: make/CompileJavaModules.gmk [hunk 5]
```diff
@@ -117,6 +116,50 @@ $(eval $(call SetupJavaCompilation, $(MODULE), \
...
 
+################################################################################
+# Setup compilation for value classes in the module
+# TBD: When $(DOCLINT) was included there was an NPE in JavacTypes.getOverriddenMethods
+
+# Directory and file name suffix for jar file containing value classes
+VALUECLASSES_STR := valueclasses
+
+ifneq ($(COMPILER), bootjdk)
+    MODULE_VALUECLASS_SRC_DIRS := $(call FindModuleValueClassSrcDirs, $(MODULE))
```
Comment (erikj79) [inline make/CompileJavaModules.gmk:127]: Please use 2 spaces for logic block indentation. (see https://openjdk.org/groups/build/doc/code-conventions.html)

File: make/CompileJavaModules.gmk [hunk 6]
```diff
@@ -117,6 +116,50 @@ $(eval $(call SetupJavaCompilation, $(MODULE), \
...
+    MODULE_VALUECLASS_SOURCEPATH := $(call GetModuleValueClassSrcPath)
+
+    ifneq ($(MODULE_VALUECLASS_SRC_DIRS),)
+      $(eval $(call SetupJavaCompilation, $(MODULE)-$(VALUECLASSES_STR), \
+        SMALL_JAVA := false, \
```
Comment (erikj79) [inline make/CompileJavaModules.gmk:132]: Please use 4 spaces indentation for continuation.

**Review by erikj79 — APPROVED** [review]
Except for some minor cosmetics, the build changes are looking really good.

File: src/hotspot/share/runtime/arguments.cpp [hunk 7]
```diff
@@ -2938,25 +2997,36 @@ jint Arguments::parse_each_vm_init_arg(const JavaVMInitArgs* args, bool* patch_m
...
+  } else {
+    if (allow_append) {
+      // append path to existing module entry
+      _patch_mod_prefix->at(i)->append_path(path);
+    } else {
+      if (strcmp(module_name, JAVA_BASE_NAME) == 0) {
+        vm_exit_during_initialization("Cannot specify " JAVA_BASE_NAME " more than once to --patch-module");
+      } else {
+        vm_exit_during_initialization("Cannot specify a module more than once to --patch-module", module_name);
+      }
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/arguments.cpp:3027]: This just reduces to:
```
else {
  vm_exit_during_initialization(err_msg("Cannot specify %s more than once to --patch-module", module_name));
}
```
or even:
```
vm_exit_during_initialization("Cannot specify a module more than once to --patch-module", module_name);
```

**Review by dholmes-ora — COMMENTED** [review]
Sorry for the delay (I don't get email notifications for this PR).

The hotspot changes are looking good.

One minor nit.

File: src/hotspot/share/runtime/arguments.cpp (hunk 7 above)
Comment (RogerRiggs) [reply src/hotspot/share/runtime/arguments.cpp:3027]: Right, for some reason there are several tests that expect different error messages due to the prior implementation split between hotspot argument parsing and ModuleBootstrap.
I couldn't see clearly far enough in the EA/Preview/release cycle to be sure of the final disposition and so punted on correcting everything everywhere.
Eventually, the provision to replace some classes with value classes unnecessary. The classes will become permanent.
But that can only happen after the feature is no-longer preview and is fully integrated.

**Review by mlchung — APPROVED** [review]
Hi @mlchung, thanks for making a comment in an OpenJDK project!

All comments and discussions in the OpenJDK Community must be made available under the OpenJDK [Terms of Use](https://openjdk.java.net/legal/tou/). If you already are an OpenJDK [Author](https://openjdk.java.net/bylaws#author), [Committer](https://openjdk.java.net/bylaws#committer) or [Reviewer](https://openjdk.java.net/bylaws#reviewer), please click [here](https://bugs.openjdk.java.net/secure/CreateIssue.jspa?pid=11300&issuetype=1) to open a new issue so that we can record that fact. Please Use "Add GitHub user mlchung" for the summary.

If you are not an OpenJDK Author, Committer or Reviewer, simply check the box below to accept the OpenJDK Terms of Use for your comments.

 - [ ]  I agree to the [OpenJDK Terms of Use](https://openjdk.java.net/legal/tou/) for all comments I make in a project in the [OpenJDK GitHub organization](https://github.com/openjdk).

Your comment will be automatically restored once you have accepted the OpenJDK [Terms of Use](https://openjdk.java.net/legal/tou/).

**Pull request closed without merging**